### PR TITLE
fix(mcp): report a merge the rollback could not reverse (#1450)

### LINKED ISSUE #1450: fix(mcp): rollback reports success when a merge action could not be reversed
Follow-up to #1440 / PR #1441. Found by a belated second-opinion review of the merged diff.

## The gap

`services/sleep/undo.py:revert_shadow_merge_edge()` returns `False` for **two materially different
situations** — its own docstring says so:

```
Returns:
    True if an edge was restored or deleted; False if nothing matched
    (the shadow merge was already undone, or the edge was retyped by a
    later writer).
```

1. **already undone** — benign. The rollback has nothing to do; logging and continuing is correct.
2. **edge retyped by a later writer** — *not* benign. The rollback **could not reverse a recorded
   merge action** without clobbering newer state. The `edge_type == EDGE_TYPE_SUPERSEDES` predicate
   in both the UPDATE and the DELETE branch is exactly the guard that makes this case return `False`
   rather than blindly overwriting an unrelated association (deliberate — see the docstring).

`handle_rollback_sleep_run` collapses both into one `logger.warning(...)` and continues.

## Why it matters now

For case 2 the operator-visible result is:

| field | value |
|---|---|
| `status` | success |
| `error` | *(absent — not `partial_rollback`)* |
| `rollback_summary.errors` | `[]` |
| `rollback_summary.merges_reversed` | `0` |
| report status | `rolled_back` |

Nothing but a zero counter distinguishes "rolled back cleanly" from "one recorded merge action was
left un-reversed because newer state is in the way". A stdout warning line is the only signal, and
it is not in the response at all.

**Scenario.** A shadow merge writes a `supersedes` edge (winner→loser). Another writer retypes that
edge before the rollback runs. Rollback reports success; the `supersedes` edge is still gone from
the caller's point of view but the merge was never actually reversed.

## Note on how this arose — it is a trade, not a pure regression

Before #1441, this branch raised `TypeError` (stdlib logger + structlog kwargs), which the
per-action `except` turned into `partial_rollback`. So case 2 *was* surfaced to the operator —
loudly, but with a nonsense reason (`Logger._log() got an unexpected keyword argument 'src_id'`).

#1441 restored the code's documented intent ("log and carry on"), which is correct for case 1 and
makes case 2 silent. The bug being fixed here is that **the intent itself does not distinguish the
two cases** — that predates #1441 and was simply masked by the crash.

## Suggested direction (not prescriptive)

Make `revert_shadow_merge_edge` return a tri-state instead of `bool` — e.g. `RESTORED` /
`ALREADY_UNDONE` / `BLOCKED_BY_NEWER_STATE` — and have the run-level rollback record only the third
as a rollback error (or a dedicated `merges_unreversible` counter, so it is visible without
overloading `errors`).

The helper is deliberately shared by run-level rollback and per-merge undo so the two cannot drift,
so **both call sites need to agree** on the new contract. The per-merge undo path currently returns
410 for the same state, which is worth reconciling in the same change.

## Acceptance criteria

- [ ] A rollback that cannot reverse a recorded merge action is distinguishable from a clean one in
      the **response**, not just in logs.
- [ ] "Already undone" still does not produce an error (that is the #1441 behaviour, and it is right).
- [ ] Run-level rollback and per-merge undo agree on the classification.
- [ ] Test covering the retyped-edge case specifically, not just the generic `False`.

## Provenance

The second-opinion review that surfaced this could not run its own tests (no usable temp dir in the
sandbox), so this is a static finding — the conflated-`False` claim is verified against the
docstring and the two SQL predicates, but the end-to-end scenario has not been reproduced.

### PR DESCRIPTION
**Pull request by JFK**
Closes #1450

## The bug

`revert_shadow_merge_edge` returned `bool`, and `False` meant two different
things — its own docstring said so:

1. **already undone** — benign, nothing to do
2. **a later writer changed the edge** — the recorded merge was **not reversed**

Both callers collapsed them. For case 2 the run-level rollback answered:

| field | value |
|---|---|
| `status` | success |
| `error` | *absent* |
| `rollback_summary.errors` | `[]` |
| `rollback_summary.merges_reversed` | `0` |

A zero counter was the only thing separating "rolled back cleanly" from "one
recorded action was left un-reversed", and the real signal never left stdout.

## Why it was invisible until now

Before #1441 this branch crashed (stdlib logger + structlog kwargs) and case 2
surfaced as `partial_rollback` — loudly, for a nonsense reason. #1441 restored
the intended log-and-continue, correct for case 1, silent for case 2. **The
intent itself never distinguished them**; the crash was masking that.

## The fix

`ShadowEdgeRevert` StrEnum — `RESTORED` / `ALREADY_UNDONE` /
`BLOCKED_BY_NEWER_STATE` — so neither caller can re-conflate them.

On a zero-row match the edge is read back to classify. `unique_edge` is keyed on
`(user, src, dst)`, so there is at most one row:

| | no edge found | edge of the snapshot's type | any other edge |
|---|---|---|---|
| **with `prior_edge`** | BLOCKED | ALREADY_UNDONE | BLOCKED |
| **no snapshot** (merge created it) | ALREADY_UNDONE | — | BLOCKED |

With a snapshot the undo target *is* the snapshot's type, which is what tells a
completed undo from an overwrite. Without one the target is "no edge", so a
surviving edge is a later writer's association this undo must not delete.

### Callers

- **run-level rollback** — `BLOCKED` increments a new `merges_unreversible`
  counter **and** appends to `errors`, so the run reports `partial_rollback`.
  That is exactly what its sibling case (a merge loser that could not be
  restored) has always done. The counter alone would leave callers who check
  only top-level `status`/`error` in the original silent-success hole.
  `ALREADY_UNDONE` keeps #1441's log-and-carry-on.
- **per-merge undo** — `BLOCKED` raises `edge_changed` instead of
  `already_restored`. Both stay **409** (the request conflicts with current state
  either way); the `error_code` is what says whether anything needs doing.

The MCP tool description declares the new counter — "visible in the response" is
the whole point.

## A wording correction from review

The first draft said the blocked case meant "the merge is still in effect".
**That is false.** Reaching a blocked state means the guarded statement found
*no* `supersedes` edge, so nothing is shadowing the loser in *any* blocked case.
What is actually un-reversed is the edge state the merge overwrote.

This also fixed a naming hole: `edge_retyped` did not cover a snapshot that
cannot be restored because a later writer **deleted** the edge outright. Renamed
to `edge_changed`, with a test asserting the message does not make the old claim.

Also from review: the endpoint's OpenAPI `409` described only two of the three
conflict reasons — now all three, with their error codes.

## Tests

`tests/services/sleep/test_undo_merge.py` + `tests/mcp_server/test_sleep_tools.py`
(718 passing across `tests/services/sleep/` and `tests/mcp_server/`):

- no snapshot: edge gone → `already_restored`; edge survives → `edge_changed`
- with snapshot: back at snapshot type → `already_restored`; other type →
  `edge_changed`; deleted → `edge_changed` **and** no "still in effect" claim
- run-level `BLOCKED` → `merges_unreversible == 1`, message in
  `rollback_summary.errors`, `error == "partial_rollback"`
- #1440's regression guard kept — its stub moved `False` → `ALREADY_UNDONE`,
  which is the branch that still makes the structlog-kwargs `logger.warning`
  call, so it guards the same thing

`ruff` clean, `pyright` clean on all three changed modules.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: backend/src/services/sleep/undo.py [hunk 1]
```diff
@@ -52,10 +82,14 @@ async def revert_shadow_merge_edge(
...
+    tell the two zero-row causes apart (#1450). ``unique_edge`` is keyed on
+    (user, src, dst), so there is at most one row to inspect.
+
     Returns:
-        True if an edge was restored or deleted; False if nothing matched
-        (the shadow merge was already undone, or the edge was retyped by a
-        later writer).
+        A :class:`ShadowEdgeRevert`. Callers MUST treat
+        ``BLOCKED_BY_NEWER_STATE`` as a failure to reverse — it is the case
+        where the merge is still in effect.
```
[the remark this answers is not in the document]
Comment (JFK) [reply backend/src/services/sleep/undo.py]: Right, and thank you — I corrected the operator-facing message in the previous commit but left the same false claim in the helper's own Returns docstring and two test docstrings. All three are gone in 96d35820; the only surviving occurrence is the negative assertion that guards against it coming back.
